core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1567,24 +1518,104 @@ class KafkaApis(val requestChannel: RequestChannel,
       val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
       val shareFetchContext = sharePartitionManager.newContext(groupId, shareFetchData, forgottenTopics, topicNames, newReqMetadata)
 
+      val erroneousAndValidPartitionData : ErroneousAndValidPartitionData = shareFetchContext.getErroneousAndValidTopicIdPartitions
+      val topicIdPartitionSeq : mutable.Set[TopicIdPartition] = mutable.Set()
+      erroneousAndValidPartitionData.erroneous.forEach {
+        case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+      }
+      erroneousAndValidPartitionData.validTopicIdPartitions.forEach {
+        case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+      }
+      shareFetchData.forEach {
+        case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+      }
+
+      val authorizedTopics = authHelper.filterByAuthorized(
+        request.context,
+        READ,
+        TOPIC,
+        topicIdPartitionSeq
+      )(_.topicPartition.topic)
+
+      // Handling the Acknowledgements from the ShareFetchRequest
+
+      if (!authHelper.authorize(request.context, READ, GROUP, groupId)) {
+        shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.GROUP_AUTHORIZATION_FAILED.exception) match {
+          case response: ShareAcknowledgeResponse => response
+          case _ => null
+        }
+      } else {
+        val isAcknowledgeDataPresent = isAcknowledgeDataPresentInFetchRequest()
+        if (shareSessionEpoch == ShareFetchMetadata.INITIAL_EPOCH) {
+          if (isAcknowledgeDataPresent) {
+            // If this is a full fetch request, it should not contain any acknowledgement Data
+            shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception) match {
+              case response: ShareAcknowledgeResponse => response
+              case _ => null
+            }
+          }
+          else {
+            shareAcknowledgeResponse = ShareAcknowledgeResponse.of(
+              Errors.NONE,
+              AbstractResponse.DEFAULT_THROTTLE_TIME,
+              new util.LinkedHashMap[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData],
+              Collections.emptyList())
+          }
+        }
+        else {
+          if (!isAcknowledgeDataPresent) {
+            shareAcknowledgeResponse = ShareAcknowledgeResponse.of(
+              Errors.NONE,
+              AbstractResponse.DEFAULT_THROTTLE_TIME,
+              new util.LinkedHashMap[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData],
+              Collections.emptyList())
+          }
+          val shareSessionError: Errors = sharePartitionManager.acknowledgeShareSessionCacheUpdate(groupId, Uuid.fromString(memberId), shareSessionEpoch, true)
```

Comment:
This line is wrong. It causes a double-increment of the shareSessionEpoch which means the request fails.

Reply:
I suppose there is still the possibility of a shareSessionError here, even though the increment of the epoch that this method currently performs doesn't seem appropriate now you've re-ordered it all.